### PR TITLE
feat(@bazel/jasmine): add coverage reporting

### PR DESCRIPTION
**Pull request by Toxicable**
## PR Checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)

## What is the current behavior?
jasmine_node_test does not report code coverage

## What is the new behavior?
jasmine_node_test does reports code coverage

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

Example oputput:
```
:~/Documents/rules_nodejs/packages/jasmine$ bazel run test:coverage
INFO: Invocation ID: c4e9cd85-aad8-4fad-9537-71c8a665861c
INFO: Analysed target //test:coverage (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //test:coverage up-to-date:
  bazel-bin/test/coverage_loader.js
  bazel-bin/test/coverage.sh
INFO: Elapsed time: 0.190s, Critical Path: 0.01s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //test:coverage
-----------------------------------------------------------------------------
Randomized with seed 34724
Started
.


1 spec, 0 failures
Finished in 0.006 seconds
Randomized with seed 34724 (jasmine --random=true --seed=34724)

=============================== Coverage summary ===============================
Statements   : 0% ( 0/48 )
Branches     : 0% ( 0/20 )
Functions    : 0% ( 0/9 )
Lines        : 0% ( 0/47 )
================================================================================
```

### CODE REVIEW

**Comment by Toxicable** [comment]
While we do have it working so far I'm running into an issue where istanbul is istrumenting files that aren't apart of the sources.
Since we have a well define list of soruces that should be instrumented I think the best option would be to see if istanbul can implement an includes api rather than trying to use thei excludes api.

Without any exlucdes it'll auto discover these files, producing an incorrect sode coverage report.
```
"/home/f.wiles/Documents/rules_nodejs/third_party/github.com/buffer-from/index.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map-support/source-map-support.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/lib/array-set.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/lib/base64-vlq.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/lib/base64.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/lib/binary-search.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/lib/mapping-list.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/lib/quick-sort.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/lib/source-map-consumer.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/lib/source-map-generator.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/lib/source-node.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/lib/util.js
/home/f.wiles/Documents/rules_nodejs/third_party/github.com/source-map/source-map.js
/home/f.wiles/.cache/bazel/_bazel_f.wiles/6b62bb60168067291695fe6dd86d865d/external/nodejs/bin/yarnpkg/bin/yarn.js
/home/f.wiles/.cache/bazel/_bazel_f.wiles/6b62bb60168067291695fe6dd86d865d/external/nodejs/bin/yarnpkg/lib/cli.js
/home/f.wiles/.cache/bazel/_bazel_f.wiles/6b62bb60168067291695fe6dd86d865d/external/nodejs/bin/yarnpkg/lib/v8-compile-cache.js
/home/f.wiles/Documents/rules_nodejs/packages/jasmine/src/jasmine_runner.js"
```

**Comment by alexeagle** [comment]
Seems like Istanbul hacking might not be needed anymore

> In Node.js >=10.10.0 you can now set the environment variable NODE_V8_COVERAGE to a directory, this will result in V8 coverage reports being output in this location.
https://medium.com/the-node-js-collection/rethinking-javascript-test-coverage-5726fb272949

We could try that?

**Comment by Toxicable** [comment]
Oh wow, that looks awesome.
Yeah we'll have a look into that and see how we can intergrate it with the test runner. 

It does mean we'll need to push a requirement on the nodejs version.
Could be done as a warning or maybe a flag to enable coverage collection + an error if the version isn't right.

**Comment by Toxicable** [comment]
@alexeagle  Started inital work on v8-coverage.
the way im planning on doing it is by adding another .js file to the jasmine rule: `jasmine_runner_host.js`, this file will be the new entry point and will fork itself so it can set the `NODE_V8_COVERAGE` flag to a known directory for processing after tests and been executed.

However im running into issues where the sub process (jasmine_runner.js) is unable ro resolve "bazel" nodejs paths ie: `Cannot find module 'npm_bazel_jasmine/test/foo_test.js'` any ideas why this might happen?
Or if there's a better way to set the environment variable to a dir we know and and work on later on

**Comment by Toxicable** [comment]
I think this is ready for an inital review now

From looking at https://github.com/bazelbuild/rules_nodejs/issues/312
I was able to execute the forked script, but it feels pretty hacky and I'm not sure if it'll work in other situations.

```
Executing tests from //test:coverage_test
-----------------------------------------------------------------------------
Randomized with seed 23022
Started
.


1 spec, 0 failures
Finished in 0.014 seconds
Randomized with seed 23022 (jasmine --random=true --seed=23022)

=============================== Coverage summary ===============================
Statements   : 90% ( 9/10 )
Branches     : 60% ( 3/5 )
Functions    : 100% ( 1/1 )
Lines        : 90% ( 9/10 )
================================================================================
```

**Comment by Toxicable** [comment]
So looks like the fork script hack only works when you run the test with `bazel run` if you use bazel test it fails to resolve the path.
Any advice how I can hook up these paths?

**Comment by alexeagle** [comment]
I don't see why you need to fork a process. We have control over the nodejs options passed to the original nodejs process that starts the tests. Can we set the flag there?

**Comment by Toxicable** [comment]
@alexeagle Sure, all we need is to set an environment variable.
I couldn't find how to do that with a macro setup, meaning, at a guess we'd need to translate this into a Rule rather than a Macro?
Also this was the easiest way I knew just to see if it would be possible - which we proved true

**Comment by Toxicable** [comment]
Quick update here:

I tried doing this without using a forked script but it would appear the coverage file is only written once the node process exits. Meaning this must be done from some other process somehow.

**Comment by Toxicable** [comment]
Ignore what I said above, turns out I was all wrong, you can produce coverage without a new process.
I switched to a new lib: `v8-coverage` which internally starts the Nodejs inspector module.
Can be previewed with `bazel run test:coverage_test`, gives the output below.

```
Fabians-MacBook-Air:jasmine fabianwiles$ bazel run test:coverage_test
...
Executing tests from //test:coverage_test
-----------------------------------------------------------------------------
Debugger listening on ws://127.0.0.1:57032/c20d1245-cce8-42e5-b956-437c8b21dddf
For help, see: https://nodejs.org/en/docs/inspector
Randomized with seed 12938
Started
.
1 spec, 0 failures
Finished in 0.008 seconds
Randomized with seed 12938 (jasmine --random=true --seed=12938)

=============================== Coverage summary ===============================
Statements   : 80% ( 8/10 )
Branches     : 75% ( 3/4 )
Functions    : 100% ( 0/0 )
Lines        : 80% ( 8/10 )
================================================================================
```

**Comment by Globegitter** [comment]
@toxicable if you want to be able to set the env variable before the target starts there is a generic solution, you can use the `command` rule from here: https://github.com/atlassian/bazel-tools/blob/master/multirun/README.md

Plus I also opened an issue: https://github.com/bazelbuild/bazel/issues/7364 as I (and some of our devs on the team) expected this to be part of bazel in the same way as one can set args generically.

**Comment by Globegitter** [comment]
Also are there any plans to integrate this with the `bazel coverage` command?

**Comment by Toxicable** [comment]
@Globegitter Ahhh, thanks for the heads up on the `command` rule.
However, for this situation we'd still need to run some JS after the `jasmine_runner` has exited, since it would appear that using  the `V8_COVERAGE` environment variable only writes the coverage once it exits the process. So we would still have to run some JS host that forks into `jasmine_runner`.
I think the current solution should work, aslong as there's no weird interactions with other apis due to using the inspector api.

As for  `bazel coverage` - I'd definitely like to integrate with it and from here it should easy to do so really since we can go to any coverage format, I just don't know how, I havn't been able to find any docs on how to do so.
So for now using `text-summary` will work as a stop gap, I'm not sure if we want to merge it in this state without `bazel coverage` integration though.

**Comment by alexeagle** [comment]
I'll discuss with the Bazel team this week about adding proper `bazel coverage` command. I'll leave it to @Globegitter and @gregmagolan about whether to merge this with a text summary or wait for guidance.

File: packages/jasmine/src/jasmine_node_test.bzl [hunk 1]
```diff
@@ -64,6 +78,7 @@ def jasmine_node_test(
         templated_args = ["$(location :%s_devmode_srcs.MF)" % name],
         testonly = 1,
         expected_exit_code = expected_exit_code,
+        configuration_env_vars = ["NODE_V8_COVERAGE"],
```
Comment (Globegitter) [inline packages/jasmine/src/jasmine_node_test.bzl]: Why do we expect this to be passed in by the user?

File: .bazelrc [hunk 2]
```diff
@@ -1,12 +0,0 @@
-# Print test logs for failed tests
```
Comment (Globegitter) [inline .bazelrc]: Why is this removed?

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (Globegitter) [inline .vscode/settings.json]: Maybe we can add this to gitignore

File: packages/jasmine/src/jasmine_runner.js [hunk 4]
```diff
@@ -27,6 +33,9 @@ const BAZEL_EXIT_NO_TESTS_FOUND = 4;
 // the maximum (See: https://nodejs.org/api/errors.html#errors_error_stacktracelimit)
 Error.stackTraceLimit = Infinity;
 
+const IS_TEST_FILE = /[^a-zA-Z0-9](spec|test)\.js$/i;
+const IS_MODULE_MODULE = /\/node_modules\//
```
Comment (Globegitter) [inline packages/jasmine/src/jasmine_runner.js]: Did you mean to call it this, or maybe a typo?

Comment (Globegitter) [reply packages/jasmine/src/jasmine_runner.js]: The variable name I mean btw.

File: packages/jasmine/src/jasmine_node_test.bzl [hunk 5]
```diff
@@ -51,8 +51,22 @@ def jasmine_node_test(
         tags = tags,
     )
 
-    all_data = data + srcs + deps + [jasmine]
-    all_data += [Label("//:src/jasmine_runner.js")]
+    runner_name = name + "." + "runner"
+
+    nodejs_binary(
```
Comment (Globegitter) [inline packages/jasmine/src/jasmine_node_test.bzl]: Any particular reason to wrap this in a `nodejs_binary` compared to how it was before?

File: packages/jasmine/BUILD.bazel [hunk 6]
```diff
@@ -18,6 +18,7 @@ package(default_visibility = ["//visibility:public"])
 
 exports_files([
     "src/jasmine_runner.js",
+    "src/jasmine_runner_host.js",
```
Comment (Globegitter) [inline packages/jasmine/BUILD.bazel]: I don't see this file. Leftover?

**Comment by Toxicable** [comment]
@Globegitter Thanks for the review. Will do a tidy up tonight.

Right now it's on no matter what. There is a small perf overhead ~20% with this.
Maybe we should provide an attribute to opt-in incase there's issues we havn't seen with it so far, then later on make a breaking change to switch to opt-out to encourage people to use the coverage report.

Im not sure what you mean by 
>regex for finding the coverage files to be configurable,

This is the part where we filter out the coverage collected from external source files
https://github.com/bazelbuild/rules_nodejs/pull/553/files#diff-b7f7068522dfd058878f39fba971463aR87
which is just a direct file match, I don't see why you'd want it to be configurable though, this will always pick up all user source files that were in the test, is there case where you'd only want a subset of that?

**Comment by Globegitter** [comment]
Yeah I think we should have coverage turned off by default for now and then later we can switch it on. One way I can see it work is by adding a `coverage` parameter to the jasmine_node_test macro and if that is set you can pass a custom arg via `args` to the nodejs_test and then that can be checked in the runner.

**Comment by Globegitter** [comment]
About the file match, yeah the static one you have works for idiomatic bazel usage, but I was thinking of code bases like our own where we have common js code in sloppy file group targets that have a bunch of files grouped together even if an app might just need a subset. So in that case we might want to be able to specify our own regex to filter out specific files. But as I said I think it is good to keep it simple and promote idiomatic use. So I think it makes sense to just leave it as is.

**Comment by Toxicable** [comment]
@Globegitter Ah yes, I see your point, makes sense.

I've implemented the coverage flag and disabled it by default.
Your arguments idea works well here
You can re-enable it with `coverage = True` in your `jasmine_node_test`

One remaining point is that I'm using the dep `"@npm//v8-coverage"` - should we make a `npm_bazel_jasmine_deps` yarn_install to get this dep from instead?
I've noticed there's been a bit of change about where these deps are coming from lately so not sure on the best option.

File: packages/jasmine/src/jasmine_node_test.bzl [hunk 7]
```diff
@@ -51,6 +53,13 @@ def jasmine_node_test(
         tags = tags,
     )
 
+    enableCoverage = False
```
Comment (Globegitter) [inline packages/jasmine/src/jasmine_node_test.bzl]: no need for the `enableCoverage` flag, we can just use coverage directly below.

**Comment by Globegitter** [comment]
I am also not familiar with the new process yet, but as far as I understand, once this is merged a new npm package has to be built and then one just adds the latest `@bazel/jasmine` to their package.json and will get the dependencies via normal npm resoultion. Maybe @alexeagle or @gregmagolan can fill in on how the release / dependency process is currently working.

**Comment by gregmagolan** [comment]
Yes @Globegitter you are correct. `/packages/jasmine` gets packaged up by `//packages/jasmine:npm_package` and released as `@bazel/jasmine`. That npm package installs the `npm_bazel_jasmine` workspace via the generated `@npm://install_bazel_workspaces.bzl` and it also pulls in transitive npm deps such as the `jasmine` npm package.

**Comment by gregmagolan** [comment]
For CI & developing & testing locally I made a number shell scripts that build the packages before running the e2e test and examples. You can call `./scripts/build_packages.sh jasmine` to build the jasmine package and its is copied out of the `bazel-out` folder into `/dist/jasmine`. The e2e tests and examples then link to this package in their package.json.

**Comment by gregmagolan** [comment]
So, if `@bazel/jasmine` requires the `v8-coverage` npm package then that just needs to be added as a dependency to `/packages/jasmine/package.json` and it will be pulled into a user's node_modules transitively when the install `@bazel/jasmine`.

Hmmmm. I think I need to spend some time to document this in `DEVELOPING.md` :)

File: packages/jasmine/src/jasmine_node_test.bzl (hunk 7 above)
Comment (Toxicable) [reply packages/jasmine/src/jasmine_node_test.bzl]: Ah, does `1` get converted to a boolean?
I've seen in some places `testonly = 1` and `testonly = True` being treated as equivilent

Comment (Toxicable) [reply packages/jasmine/src/jasmine_node_test.bzl]: it would appear not, I think it's reasonable to ask for just `True` or `False` as values here though

**Comment by Toxicable** [comment]
Alrighty, I think we're good to go here now.
I added a bit of a summary to the commit message description of this feature and it's current status.
Also debugging still appears to work even with the internal call the inspector API

File: packages/jasmine/src/jasmine_node_test.bzl [hunk 8]
```diff
@@ -51,7 +55,13 @@ def jasmine_node_test(
         tags = tags,
     )
 
+    # append on any user args to the end
+    args = ["--coverage=%s" % coverage] + args
+
     all_data = data + srcs + deps + [jasmine]
+
+    # code coverage reporting dependancy
+    all_data += ["@npm//v8-coverage"]
```
Comment (gregmagolan) [inline packages/jasmine/src/jasmine_node_test.bzl]: You can remove this line. Pulling in `@npm//@bazel/jasmine` (the `jasmine` attribute) is sufficient to also pull in `v8-coverage` since its a transitive dep. Also, we don't want to assume `@npm` is the users node_modules workspace. That's why we have the `jasmine` attribute so a user can set it to `@foobar//@bazel/jasmine` if necessary.

File: packages/jasmine/test/BUILD.bazel [hunk 9]
```diff
@@ -30,3 +30,13 @@ jasmine_node_test(
...
 )
+
+jasmine_node_test(
+    name = "coverage_test",
+    srcs = [
+        "coverage.spec.js",
+        "coverage_source.js",
+    ],
+    coverage = True,
+    jasmine = "@npm//jasmine",
```
Comment (gregmagolan) [inline packages/jasmine/test/BUILD.bazel]: will need to rebase and add `deps = ["//:jasmine_runner"],` to this target

File: packages/jasmine/src/jasmine_runner.js [hunk 10]
```diff
@@ -138,5 +207,5 @@ function getAllSpecs(jasmineEnv) {
 }
 
 if (require.main === module) {
-  process.exitCode = main(process.argv.slice(2));
+    process.exitCode = main(process.argv.slice(2));
```
Comment (gregmagolan) [inline packages/jasmine/src/jasmine_runner.js]: nit: extra whitespace

File: packages/jasmine/src/jasmine_node_test.bzl [hunk 11]
```diff
@@ -51,7 +55,13 @@ def jasmine_node_test(
         tags = tags,
     )
 
+    # append on any user args to the end
+    args = ["--coverage=%s" % coverage] + args
```
Comment (gregmagolan) [inline packages/jasmine/src/jasmine_node_test.bzl]: You should be able to just add this argument to `templates_args` and avoid the `args` altogether. Also, looking for `--coverage=True` seems pretty brittle as a user could set it to `1`. This would be better:

```
templated_args = ["$(location :%s_devmode_srcs.MF)" % name]
if coverage:
  templated_args = templated_args + ["--coverage"]
else:
  templated_args = templated_args + ["--nocoverage"]
```

and then just look for an argument matching `--coverage` in the `jasmine_runner.js`

File: packages/jasmine/test/BUILD.bazel (hunk 9 above)
Comment (gregmagolan) [reply packages/jasmine/test/BUILD.bazel]: Can you also add a smoke test to `/e2e/jasmine`? It would test the `v8-coverage` can be required in user land when using the `@bazel/jasmine` package.
A test which verifies the report works would be good but I think we can merge this without one and add that in a follow up PR.

File: packages/jasmine/src/jasmine_runner.js [hunk 12]
```diff
@@ -1,7 +1,13 @@
 const fs = require('fs');
+const path = require('path');
+const crypto = require('crypto');
```
Comment (gregmagolan) [inline packages/jasmine/src/jasmine_runner.js]: crypo & Execute & Report requires can be done within the `if (enableCoverage)` blocks so in the no coverage case they are no required at all

File: packages/jasmine/test/coverage_source.js [hunk 13]
```diff
@@ -0,0 +1,9 @@
+function isString(input) {
+    if(typeof input === 'string') {
```
Comment (gregmagolan) [inline packages/jasmine/test/coverage_source.js]: nit: `if (`

File: packages/jasmine/src/jasmine_node_test.bzl (hunk 8 above)
Comment (Toxicable) [reply packages/jasmine/src/jasmine_node_test.bzl]: I'm running into issues with this one.
If I remove it then im not able to run the tests inside the jasmine package anymore, since nothing is pulling `@npm//v8-coverage` anymore

Comment (Globegitter) [reply packages/jasmine/src/jasmine_node_test.bzl]: Yeah I would assume it is just an issue of getting the right package locally now so it pulls in the right deps automatically. I will take a look myself if I spot anything obvious.

Comment (gregmagolan) [reply packages/jasmine/src/jasmine_node_test.bzl]: I think you just need to add `@npm//v8-coverage` explicitly to the deps (or data) of the test within `packages/jasmine/test/BUILD.bazel`:
```
jasmine_node_test(
    name = "coverage_test",
    srcs = [
        "coverage.spec.js",
        "coverage_source.js",
    ],
    coverage = True,
    jasmine = "@npm//jasmine",
    deps = [
        "//:jasmine_runner",
        "@npm//v8-coverage,
    ],
)
```

Comment (Toxicable) [reply packages/jasmine/src/jasmine_node_test.bzl]: ah! easy as, sorted

File: packages/jasmine/test/BUILD.bazel (hunk 9 above)
Comment (Toxicable) [reply packages/jasmine/test/BUILD.bazel]: done

**Comment by Toxicable** [comment]
@gregmagolan Thanks for the review, I've addressed all your points, let me know if there's anything else.
With the exception of adding an e2e test for the text-summary output. Since in my next PR ill be turning that off `text-summary` and going to lcov in an attempt to intergrade with bazel coverage

**Comment by Globegitter** [comment]
@gregmagolan @Toxicable  One thing I forgot is documentation, I assume skydoc should pick up the added comment about coverage automatically, but is there anything automatically updating the docs? Looking at https://bazelbuild.github.io/rules_nodejs/jasmine_node_test/jasmine_node_test.html#jasmine_node_test I can not see the new `coverage` attribute commented.

**Comment by gregmagolan** [comment]
> @gregmagolan @Toxicable One thing I forgot is documentation, I assume skydoc should pick up the added comment about coverage automatically, but is there anything automatically updating the docs? Looking at https://bazelbuild.github.io/rules_nodejs/jasmine_node_test/jasmine_node_test.html#jasmine_node_test I can not see the new `coverage` attribute commented.

We typically update the docs with each release so the docstring from jasmine_node_test.bzl will get updated at that point:

```
def jasmine_node_test(
        ...
        **kwargs):
    """Runs tests in NodeJS using the Jasmine test runner.

    To debug the test, see debugging notes in `nodejs_test`.

    Args:
      name: name of the resulting label
      srcs: JavaScript source files containing Jasmine specs
      data: Runtime dependencies which will be loaded while the test executes
      deps: Other targets which produce JavaScript, such as ts_library
      expected_exit_code: The expected exit code for the test. Defaults to 0.
      tags: bazel tags applied to test
      jasmine: a label providing the jasmine dependency
      coverage: Enables code coverage collection and reporting
      **kwargs: remaining arguments are passed to the test rule
    """
```

which includes the short description of `coverage`.

@alexeagle Has done more work on docs and some things have changed since moving to separate npm packages such as `@npm//jasmine`. I don't know if we have the docs from each package hosted at `bazelbuild.github.io` yet. The docs there may still be from the legacy `/internal/jasmine_node_test` rule.
